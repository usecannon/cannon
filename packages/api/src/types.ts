import type { Address } from 'viem';

export type IpfsUrl = `ipfs://${string}`;

export interface ApiNamespace {
  type: 'namespace';
  name: string;
  count: number;
}

export interface ApiPackage {
  type: 'package';
  name: string;
  publisher: string;
  version: string;
  preset: string;
  chainId: number;
  deployUrl: IpfsUrl;
  metaUrl: IpfsUrl;
  miscUrl: IpfsUrl;
  timestamp: number;
}

export interface ApiContract {
  type: 'contract';
  address: Address;
  contractName: string;
  name: string;
  chainId: number;
  preset: string;
  version: string;
}

type ExtractType<T> = T extends { type: infer U } ? U : never;

export type ApiDocument = ApiNamespace | ApiPackage;
export type ApiDocumentType = ExtractType<ApiDocument>;
