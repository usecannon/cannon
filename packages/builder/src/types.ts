import * as viem from 'viem';
import { Abi, Address, Hash, Hex, SendTransactionParameters } from 'viem';

import _ from 'lodash';

import type { RawChainDefinition } from './actions';

// loosely based on the hardhat `Artifact` type
export type ContractArtifact = {
  contractName: string;
  sourceName: string;
  abi: Abi;
  bytecode: Hex;
  deployedBytecode: string;
  linkReferences: {
    [fileName: string]: {
      [contractName: string]: {
        start: number;
        length: number;
      }[];
    };
  };
  source?: {
    solcVersion: string;
    input: string;
  };
};

export type ContractData = {
  address: Address;
  abi: Abi;
  constructorArgs?: any[]; // only needed for external verification
  linkedLibraries?: { [sourceName: string]: { [libName: string]: string } }; // only needed for external verification
  deployTxnHash: string;
  deployTxnBlockNumber?: string;
  deployTimestamp?: string;
  contractName: string;
  sourceName: string;
  deployedOn: string;
  highlight?: boolean;
  gasUsed: number;
  gasCost: string;
};

export type ContractMap = {
  [label: string]: ContractData;
};

export type TransactionMap = {
  [label: string]: {
    hash: Hash | '';
    blockNumber?: string;
    timestamp?: string;
    events: EventMap;
    deployedOn: string;
    gasUsed: number;
    gasCost: string;
    signer: string;
  };
};

export type EventMap = {
  [name: string]: {
    args: string[];
  }[];
};

export interface PreChainBuilderContext {
  chainId: number;

  package: any;

  timestamp: string;

  overrideSettings: { [label: string]: string };
}

export interface ChainBuilderContext extends PreChainBuilderContext {
  contracts: ContractMap;

  txns: TransactionMap;

  settings: { [label: string]: string };

  // Legacy
  extras?: { [label: string]: string };

  imports: BundledChainBuilderOutputs;

  [shortContract: string]: any;
}

const etherUnitNames = ['wei', 'kwei', 'mwei', 'gwei', 'szabo', 'finney', 'ether'];

export const CannonHelperContext = {
  // ethers style constants
  AddressZero: viem.zeroAddress,
  HashZero: viem.zeroHash,
  MaxUint256: viem.maxUint256,

  // ethers style utils
  defaultAbiCoder: {
    encode: (a: string[], v: any[]) => {
      return viem.encodeAbiParameters(
        a.map((arg) => ({ type: arg })),
        v
      );
    },
    decode: (a: string[], v: viem.Hex | viem.ByteArray) => {
      return viem.decodeAbiParameters(
        a.map((arg) => ({ type: arg })),
        v
      );
    },
  },

  zeroPad: (a: viem.Hex, s: number) => viem.padHex(a, { size: s }),
  hexZeroPad: (a: viem.Hex, s: number) => viem.padHex(a, { size: s }),
  hexlify: viem.toHex,
  stripZeros: viem.trim,
  formatBytes32String: (v: string) => viem.stringToHex(v, { size: 32 }),
  parseBytes32String: (v: viem.Hex) => viem.hexToString(v, { size: 32 }),
  id: (v: string) => (v.startsWith('function ') ? viem.toFunctionSelector(v) : viem.keccak256(viem.toHex(v))),
  formatEther: viem.formatEther,
  formatUnits: (s: bigint, units: number | string) => {
    if (typeof units === 'string') {
      const index = etherUnitNames.indexOf(units);
      if (index < 0) {
        throw new Error(`formatUnits: unknown ethereum unit name: ${units}`);
      }
      units = 3 * index;
    }

    return viem.formatUnits(s, units as number);
  },
  parseEther: viem.parseEther,
  parseUnits: (s: string, units: number | string) => {
    if (typeof units === 'string') {
      const index = etherUnitNames.indexOf(units);
      if (index < 0) {
        throw new Error(`parseUnits: unknown ethereum unit name: ${units}`);
      }
      units = 3 * index;
    }

    return viem.parseUnits(s, units as number);
  },
  keccak256: viem.keccak256,
  sha256: viem.sha256,
  ripemd160: viem.ripemd160,
  solidityPack: viem.encodePacked,
  solidityKeccak256: (a: string[], v: any[]) => viem.keccak256(viem.encodePacked(a, v)),
  soliditySha256: (a: string[], v: any[]) => viem.sha256(viem.encodePacked(a, v)),
  serializeTransaction: viem.serializeTransaction,
  parseTransaction: viem.parseTransaction,

  encodeFunctionData: viem.encodeFunctionData,
  decodeFunctionData: viem.decodeFunctionData,
  encodeFunctionResult: viem.encodeFunctionResult,
  decodeFunctionResult: viem.decodeFunctionResult,
};

export type ChainBuilderContextWithHelpers = ChainBuilderContext & typeof CannonHelperContext;

export type BuildOptions = { [val: string]: string };

export type StorageMode = 'all' | 'metadata' | 'none';

export type CannonSigner = { wallet: viem.WalletClient; address: Address };

export type Contract = Pick<viem.SimulateContractParameters, 'abi' | 'address'>;

export interface ChainBuilderRuntimeInfo {
  // Interface to which all transactions should be sent and all state queried
  provider: viem.PublicClient;

  // chainID to interact with
  chainId: number;

  // returns the signer associated with the given address. Reverts if the signer is not found or cannot be populated.
  getSigner: (addr: viem.Address) => Promise<CannonSigner>;

  // returns a signer which should be used for sending the specified transaction.
  getDefaultSigner?: (txn: Omit<SendTransactionParameters, 'account' | 'chain'>, salt?: string) => Promise<CannonSigner>;

  // returns contract information from the specified artifact name.
  getArtifact?: (name: string) => Promise<ContractArtifact>;

  // Should record snapshots?
  snapshots: boolean;

  // Should gracefully continue after failures and return a partial release?
  allowPartialDeploy: boolean;

  // Gas price to use for transactions
  gasPrice?: string;

  // Base and Priority gas fee to use for transactions - EIP1559
  gasFee?: string;
  priorityGasFee?: string;
}

export interface PackageState {
  name: string;
  version: string;
  currentLabel: string;
}

export type BundledOutput = { url: string; tags?: string[]; target?: string; preset?: string } & ChainArtifacts;

export interface BundledChainBuilderOutputs {
  [module: string]: BundledOutput;
}

export type ChainArtifacts = Partial<Pick<ChainBuilderContext, 'imports' | 'contracts' | 'txns' | 'settings' | 'extras'>>;

export interface ChainBuilderOptions {
  [key: string]: string;
}

/**
 * The deployment info is the main output of the cannon build process, and what is ultimately the blob referenced by the registry when pulling a package.
 * It contains all the information
 * needed to interface with a deployment generated by cannon.
 */
export type DeploymentInfo = {
  // used to identify that cannon created an artifact on IPFS
  generator: `cannon ${string}`;

  // the time at which this deployment was generated
  timestamp: number;

  // contents of cannonfile.toml used for this build in raw json form
  // if not included, defaults to the chain definition at the DeploymentManifest instead
  def: RawChainDefinition;

  // setting overrides used to build this chain
  options: ChainBuilderOptions;

  // defines whether this deployment has been fully completed or only partially or completely unbuilt
  status?: 'complete' | 'partial' | 'none';

  // the result of all the build steps for the last build
  state: DeploymentState;

  // additional package information. could be `package.json` file from a hardhat project, for example.
  meta: any;

  // ipfs hash additional required files for possible build
  miscUrl: string;

  // EVM chain which this deployment is for
  chainId?: number;
};

export type DeploymentManifest = {
  // contents of cannonfile.toml stringified
  def: RawChainDefinition;

  // npm style package.json for the project being uploaded
  npmPackage: any;

  // tag of the package which was used as the base for this package
  upgradeFrom?: string;

  // archive which contains miscellaneus dependencies ex. documentation pages, contracts, etc.
  misc: {
    ipfsHash: string;
  };

  deploys: {
    [chainId: string]: {
      [preset: string]: {
        hash: string;
      };
    };
  };
};

/**
 * After executing a build step, this object is recorded to the DeploymentInfo for future reference.
 */
export type StepState = {
  // The BUILD_VERSION when the step was built
  version: number;

  // A md5sum of the step inputs. Used to determine if the step should be rebuilt on upgrades.
  hash: string | null;

  // Once the step completes, the current state of the deployment artifacts is recorded
  artifacts: ChainArtifacts;

  // If this is a cannon network build, the full dump of the chain blob is recorded
  chainDump?: Hex | null; // only included if cannon network build
};

export type DeploymentState = { [label: string]: StepState };

export function combineCtx(ctxs: ChainBuilderContext[]): ChainBuilderContext {
  const ctx = _.clone(ctxs[0]);

  ctx.timestamp = Math.floor(Date.now() / 1000).toString(); //(await this.provider.getBlock(await this.provider.getBlockNumber())).timestamp.toString();

  // merge all blockchain outputs
  for (const additionalCtx of ctxs.slice(1)) {
    ctx.contracts = { ...ctx.contracts, ...additionalCtx.contracts };
    ctx.txns = { ...ctx.txns, ...additionalCtx.txns };
    ctx.imports = { ...ctx.imports, ...additionalCtx.imports };
    ctx.settings = { ...ctx.settings, ...additionalCtx.settings };
  }

  return ctx;
}
