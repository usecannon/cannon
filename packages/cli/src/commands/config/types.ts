export type Option = {
  flags: string;
  description: string;
  defaultValue?: string | boolean | string[];
};

type Argument = {
  flags: string;
  description: string;
  defaultValue?: string | boolean | string[];
};

type BaseCommand = {
  description: string;
  usage?: string;
  arguments?: Argument[];
  options?: Option[];
  anvilOptions?: Option[];
};

type SubCommand = BaseCommand;

export type Command = BaseCommand & {
  commands?: Record<string, SubCommand>;
};

type PluginCommand = Command & {
  commands: {
    list: SubCommand;
    add: SubCommand;
    remove: SubCommand;
  };
};

export type CommandsConfig = {
  [key: string]: Command;
  plugin: PluginCommand;
};
