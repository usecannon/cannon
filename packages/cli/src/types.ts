export type PackageSettings = { [k: string]: string };

export interface PackageSpecification {
  name: string;
  version: string;
  preset?: string;
  settings: PackageSettings;
}
