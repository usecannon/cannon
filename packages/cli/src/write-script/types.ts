import type { ChainArtifacts } from '@usecannon/builder';
import type { Config as StepDeploy } from '@usecannon/builder/src/steps/deploy';
import type { Config as StepInvoke } from '@usecannon/builder/src/steps/invoke';
import type { Config as StepRouter } from '@usecannon/builder/src/steps/router';
import type { Config as StepPull } from '@usecannon/builder/src/steps/pull';
import type { Config as StepClone } from '@usecannon/builder/src/steps/clone';
import * as viem from 'viem';

type BaseDumpLine = {
  label: string;
  depth: number;
  result?: ChainArtifacts; // Step is called pre step execution if missing result
  txns: viem.Transaction[]; // Executed transactions
};

/**
 * Type of each dump's line, each line has to be self contained and include all
 * the required information so it can be safely streamed during usage.
 */
export type DumpLine =
  | (BaseDumpLine & {
      type: 'deploy';
      step: StepDeploy;
    })
  | (BaseDumpLine & {
      type: 'invoke';
      step: StepInvoke;
    })
  | (BaseDumpLine & {
      type: 'router';
      step: StepRouter;
    })
  | (BaseDumpLine & {
      type: 'pull';
      step: StepPull;
    })
  | (BaseDumpLine & {
      type: 'clone';
      step: StepClone;
    });

export type DumpRenderer = () => any;
