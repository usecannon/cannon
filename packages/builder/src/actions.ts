import { z } from 'zod';

import { ChainBuilderRuntime } from './runtime';

import contractSpec from './steps/contract';

import importSpec from './steps/import';

import invokeSpec from './steps/invoke';
import keeperSpec from './steps/keeper';
import provisionSpec from './steps/provision';

import { ChainArtifacts, ChainBuilderContext, ChainBuilderContextWithHelpers, PackageState } from './types';
import { chainDefinitionSchema } from './schemas.zod';
import { handleZodErrors } from './error/zod';

export interface CannonAction {
  label: string;

  configInject: (ctx: ChainBuilderContextWithHelpers, config: any, packageState: PackageState) => any;

  getState: (
    runtime: ChainBuilderRuntime,
    ctx: ChainBuilderContextWithHelpers,
    config: any,
    packageState: PackageState
  ) => any;

  exec: (
    runtime: ChainBuilderRuntime,
    ctx: ChainBuilderContext,
    config: any,
    packageState: PackageState
  ) => Promise<ChainArtifacts>;

  // Takes in any schema as long as the base type is ZodSchema
  validate: z.ZodSchema;

  timeout?: number;
}

/**
 * @internal
 * All the different types (and their implementations)
 */
export const ActionKinds: { [label: string]: CannonAction } = {};

/**
 *  Available properties for top level config
 *  @public
 *  @group Base Cannonfile Config
 
 */
export type RawChainDefinition = z.infer<typeof chainDefinitionSchema>;

/**
 *  @internal
 *  parses the schema and performs zod validations safely with a custom error handler
 */
export function validateConfig(schema: z.ZodSchema, config: any) {
  const result = schema.safeParse(config);

  if (!result.success) {
    const errors = result.error.errors;
    handleZodErrors(errors);
  }

  return result;
}

export function registerAction(action: CannonAction) {
  if (typeof action.label !== 'string') {
    throw new Error('missing "label" property on plugin definition');
  }

  const { label } = action;

  if (ActionKinds[action.label]) {
    throw new Error('action kind already declared: ' + label);
  }

  ActionKinds[label] = action;

  (
    chainDefinitionSchema.pick({
      description: true,
      keywords: true,
      setting: true,
      import: true,
    }) as any
  )[label] = { values: action.validate };
}

registerAction(contractSpec);
registerAction(importSpec);
registerAction(invokeSpec);
registerAction(keeperSpec);
registerAction(provisionSpec);
