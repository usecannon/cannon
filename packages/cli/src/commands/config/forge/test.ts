import type { Option as ForgeTestOption } from '../types';

import { anvilOptions } from '../anvil';
import { forgeEvmOptions } from './common/evm';
import { forgeDisplayOptions } from './common/display';
import { forgeCompilerOptions } from './common/compiler';
import { forgeProjectOptions } from './common/project';

export const forgeTestOptions: ForgeTestOption[] = [
  {
    flags: '--forge.debug <test_function>',
    description: 'Run a test in the debugger.',
  },
  {
    flags: '--forge.decode-internal [<test_function>]',
    description: 'Whether to identify internal functions in traces.',
  },
  {
    flags: '--forge.gas-report',
    description: 'Print a gas report.',
  },
  {
    flags: '--forge.allow-failure',
    description: 'Exit with code 0 even if a test fails.',
  },
  {
    flags: '--forge.fail-fast',
    description: 'Stop running tests after the first failure.',
  },
  {
    flags: '--forge.etherscan-api-key <key>',
    description: 'The Etherscan (or equivalent) API key.',
  },
  {
    flags: '--forge.fuzz-seed <fuzz_seed>',
    description: 'Set seed used to generate randomness during your fuzz runs.',
  },
  {
    flags: '--forge.fuzz-runs <runs>',
    description: 'Set the number of fuzz runs.',
  },
  {
    flags: '--forge.fuzz-input-file <fuzz_input_file>',
    description: 'File to rerun fuzz failures from.',
  },
  {
    flags: '--forge.threads <threads>',
    description: 'Max concurrent threads to use. Default value is the number of available CPUs.',
  },
  {
    flags: '--forge.show-progress',
    description: 'Show test execution progress.',
  },
  {
    flags: '--forge.match-test <regex>',
    description: 'Only run test functions matching the specified regex pattern.',
  },
  {
    flags: '--forge.no-match-test <regex>',
    description: 'Only run test functions that do not match the specified regex pattern.',
  },
  {
    flags: '--forge.match-contract <regex>',
    description: 'Only run tests in contracts matching the specified regex pattern.',
  },
  {
    flags: '--forge.no-match-contract <regex>',
    description: 'Only run tests in contracts that do not match the specified regex pattern.',
  },
  {
    flags: '--forge.match-path <glob>',
    description: 'Only run tests in source files matching the specified glob pattern.',
  },
  {
    flags: '--forge.no-match-path <glob>',
    description: 'Only run tests in source files that do not match the specified glob pattern.',
  },
  {
    flags: '--forge.no-match-coverage <regex>',
    description: 'Only show coverage for files that do not match the specified regex pattern.',
  },
  {
    flags: '--forge.rerun',
    description: 'Re-run recorded test failures from last run. If no failure recorded then regular test run is performed.',
  },
  ...forgeDisplayOptions,
  ...forgeEvmOptions,
  ...forgeCompilerOptions,
  ...forgeProjectOptions,
  ...anvilOptions,
];
