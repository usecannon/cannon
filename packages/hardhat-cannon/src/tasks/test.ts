import { task } from 'hardhat/config';
import { TASK_TEST } from '../task-names';

const DEFAULT_CANNONFILE = 'cannonfile.toml';

task(TASK_TEST, 'Utility for running hardhat tests on the cannon network')
  .addOptionalPositionalParam('cannonfile', 'Path to a test cannonfile to build', DEFAULT_CANNONFILE)
  .addOptionalVariadicPositionalParam('settings', 'Custom settings for building the cannonfile', [])
  .addOptionalParam('preset', '(Optional) The preset label for storing the build with the given settings')
  .addOptionalParam(
    'anvilOptions',
    '(Optional) Custom anvil options json string or json file to configure when running on the cannon network or a local forked node'
  )
  .addOptionalParam('registryPriority', '(Optional) Which registry should be used first?', 'local')
  .addFlag('noCompile', "Don't compile before running this task")
  .addOptionalParam('testFiles', 'An optional list of files separated by a comma to test')
  .addFlag('parallel', 'Run tests in parallel')
  .addFlag('bail', 'Stop running tests after the first test failure')
  .addOptionalParam('grep', 'Only run tests matching the given string or regexp')
  .setAction(async (params, hre) => {
    await hre.run('cannon:build', params);
    return await hre.run('test', params);
  });
