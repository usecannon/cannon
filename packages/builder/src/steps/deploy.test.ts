import * as viem from 'viem';
import { fixtureTransactionReceipt } from '../../test/fixtures';
import { validateConfig } from '../actions';
import { ContractArtifact } from '../types';
import action from './deploy';
import { fakeCtx, fakeRuntime, makeFakeSigner } from './utils.test.helper';

const DEFAULT_ARACHNID_ADDRESS = '0x4e59b44847b379578588920cA78FbF26c0B4956C';

describe('steps/deploy.ts', () => {
  const fakeAbi = [
    {
      inputs: [
        {
          internalType: 'bytes32',
          name: 'data',
          type: 'bytes32',
        },
        {
          internalType: 'bytes32',
          name: 'data',
          type: 'bytes32',
        },
        {
          name: 'data',
          type: 'tuple',
          components: [
            {
              type: 'string',
              name: 'testval',
            },
          ],
        },
      ],
      stateMutability: 'nonpayable',
      type: 'constructor',
    },
  ];
  const fakeRx = fixtureTransactionReceipt();

  beforeAll(async () => {
    jest.mocked(fakeRuntime.getArtifact).mockResolvedValue({
      bytecode: '0xabcd',
      abi: fakeAbi,
    } as unknown as ContractArtifact);

    jest.mocked((fakeRuntime.provider as any).sendTransaction).mockResolvedValue('0x1234');
    jest.mocked(fakeRuntime.provider.waitForTransactionReceipt).mockResolvedValue(fakeRx);
  });

  describe('validate', () => {
    it('fails when not setting values', () => {
      expect(() => validateConfig(action.validate, {})).toThrow('Field: artifact');
    });

    it('fails when setting invalid value', () => {
      expect(() => validateConfig(action.validate, { artifact: 'Greeter', invalid: ['something'] })).toThrow(
        "Unrecognized key(s) in object: 'invalid'"
      );
    });
  });

  describe('configInject()', () => {
    it('injects all fields', async () => {
      const result = action.configInject(fakeCtx, {
        artifact: '<%= settings.a %><%= settings.b %>',

        from: '<%= settings.c %>',
        nonce: '<%= settings.d %>',
        abi: '<%= settings.c %><%= settings.d %>',
        args: ['<%= settings.d %><%= settings.a %>', { abc: '<%= settings.a %><%= settings.b %><%= settings.c %>' }],
        libraries: { dcba: '<%= settings.a %>' },

        // used to force new copy of a contract (not actually used)
        salt: '<%= settings.a %><%= settings.c %>',

        value: '<%= settings.b %><%= settings.d %>',
        overrides: {
          gasLimit: '<%= settings.gasLimit %>',
        },
      });

      expect(result).toStrictEqual({
        artifact: 'ab',

        from: 'c',
        nonce: 'd',
        abi: 'cd',
        args: ['da', { abc: 'abc' }],
        libraries: { dcba: 'a' },

        // used to force new copy of a contract (not actually used)
        salt: 'ac',

        value: 'bd',
        overrides: {
          gasLimit: '20000',
        },
      });
    });
  });

  describe('getState()', () => {
    it('resolves correct properties with minimal config', async () => {
      const result = await action.getState(fakeRuntime, fakeCtx, { artifact: 'hello' });

      expect(result).toContainEqual({
        bytecode: '0xabcd',
        args: [],
        salt: undefined,
        value: [], // for cannon spec, undefined or 0 value contract create resolves to empty array (dont ask)
      });
    });

    it('resolves correct even if there are libraries in the contract', async () => {
      // todo
    });

    it('resolves correct properties with maximal config', async () => {
      const result = await action.getState(fakeRuntime, fakeCtx, {
        artifact: 'hello',
        args: ['one', 'two', { three: 'four' }],
        salt: 'wohoo',
        value: '1234',
      });

      expect(result).toContainEqual({
        bytecode: '0xabcd',
        args: ['one', 'two', '{"three":"four"}'],
        salt: 'wohoo',
        value: '1234',
      });
    });
  });

  describe('getInputs()', () => {
    it('detects all usages', async () => {
      expect(
        action
          .getInputs(
            {
              artifact: '<%= contracts.a %>',
              abi: '<%= contracts.b %>',
              from: '<%= contracts.c %>',
              nonce: '<%= contracts.d %>',
              value: '<%= contracts.e %>',
              abiOf: ['<%= contracts.f %>', '<%= contracts.g %>'],
              args: ['<%= contracts.h %>', '<%= contracts.i %>'],
              salt: '<%= contracts.j %>',
            },
            []
          )
          .accesses.sort()
      ).toEqual([
        'contracts.a',
        'contracts.b',
        'contracts.c',
        'contracts.d',
        'contracts.e',
        'contracts.f',
        'contracts.g',
        'contracts.h',
        'contracts.i',
        'contracts.j',
      ]);
    });
  });

  describe('getOutputs()', () => {
    it('returns the contract that is outputted', () => {
      expect(action.getOutputs({ artifact: 'hello' }, { name: '', version: '', currentLabel: 'contract.Hello' })).toEqual([
        'contracts.Hello',
        'Hello',
      ]);
    });
  });

  describe('exec()', () => {
    describe('when create2 = true', () => {
      it('works if contract already deployed', async () => {
        jest.mocked(fakeRuntime.provider.getBytecode).mockResolvedValue('0xabcdef');

        const result = await action.exec(
          fakeRuntime,
          fakeCtx,
          {
            artifact: 'hello',
            create2: true,
            args: [viem.stringToHex('one', { size: 32 }), viem.stringToHex('two', { size: 32 }), { three: 'four' }],
            salt: 'wohoo',
            value: '1234',
          },
          { name: 'hello', version: '1.0.0', currentLabel: 'contract.Woot' }
        );

        expect(result).toStrictEqual({
          contracts: {
            Woot: {
              abi: fakeAbi,
              address: '0x3F9270CE7b8704E7BE0BfcA0EA8836f2B135a4ef',
              constructorArgs: [
                viem.stringToHex('one', { size: 32 }),
                viem.stringToHex('two', { size: 32 }),
                { three: 'four' },
              ],
              contractName: undefined,
              deployTxnHash: '',
              deployTxnBlockNumber: '',
              deployTimestamp: '',
              deployedOn: 'contract.Woot',
              gasCost: '0',
              gasUsed: 0,
              linkedLibraries: {},
              sourceName: undefined,
              highlight: undefined,
            },
          },
        });
      });

      it('works if contract needs to be deployed', async () => {
        jest.mocked(fakeRuntime.provider.getBytecode).mockImplementation(async ({ address }) => {
          if (address === DEFAULT_ARACHNID_ADDRESS) {
            return '0xabcd';
          }

          return '0x';
        });

        const result = await action.exec(
          fakeRuntime,
          fakeCtx,
          {
            artifact: 'hello',
            create2: true,
            args: [viem.stringToHex('one', { size: 32 }), viem.stringToHex('two', { size: 32 }), { three: 'four' }],
            salt: 'wohoo',
            value: '1234',
          },
          { name: 'hello', version: '1.0.0', currentLabel: 'contract.Woot' }
        );

        expect(result).toStrictEqual({
          contracts: {
            Woot: {
              abi: fakeAbi,
              address: '0x3F9270CE7b8704E7BE0BfcA0EA8836f2B135a4ef',
              constructorArgs: [
                viem.stringToHex('one', { size: 32 }),
                viem.stringToHex('two', { size: 32 }),
                { three: 'four' },
              ],
              contractName: undefined,
              deployTxnHash: fakeRx.transactionHash,
              deployTxnBlockNumber: '0',
              deployTimestamp: '',
              deployedOn: 'contract.Woot',
              linkedLibraries: {},
              sourceName: undefined,
              highlight: undefined,
              gasCost: '5678',
              gasUsed: 1234,
            },
          },
        });

        expect((await fakeRuntime.getDefaultSigner({}, '')).wallet.sendTransaction).toHaveBeenCalled();
      });
    });

    describe('when create2 = false', () => {
      it('deploys with specified nonce', async () => {
        const result = await action.exec(
          fakeRuntime,
          fakeCtx,
          {
            artifact: 'hello',
            highlight: true,
            args: [viem.stringToHex('one', { size: 32 }), viem.stringToHex('two', { size: 32 }), { three: 'four' }],
            salt: 'wohoo',
            value: '1234',
          },
          { name: 'hello', version: '1.0.0', currentLabel: 'contract.Woot' }
        );

        expect(result).toStrictEqual({
          contracts: {
            Woot: {
              abi: fakeAbi,
              address: fakeRx.contractAddress,
              constructorArgs: [
                viem.stringToHex('one', { size: 32 }),
                viem.stringToHex('two', { size: 32 }),
                { three: 'four' },
              ],
              contractName: undefined,
              deployTxnHash: fakeRx.transactionHash,
              deployTxnBlockNumber: '0',
              deployTimestamp: '',
              deployedOn: 'contract.Woot',
              linkedLibraries: {},
              sourceName: undefined,
              highlight: true,
              gasCost: '5678',
              gasUsed: 1234,
            },
          },
        });
      });

      it('deploys with specified signer fromCall', async () => {
        (fakeRuntime.getSigner as any) = async (addr: string) => {
          if (addr == '0x1234123412341234123412341234123412341234') {
            return makeFakeSigner('0x1234123412341234123412341234123412341234', fakeRuntime.provider as any);
          }

          return null;
        };

        const result = await action.exec(
          fakeRuntime,
          fakeCtx,
          {
            artifact: 'hello',
            from: '0x1234123412341234123412341234123412341234',
            args: [viem.stringToHex('one', { size: 32 }), viem.stringToHex('two', { size: 32 }), { three: 'four' }],
            salt: 'wohoo',
            value: '1234',
          },
          { name: 'hello', version: '1.0.0', currentLabel: 'contract.Woot' }
        );

        expect(result).toStrictEqual({
          contracts: {
            Woot: {
              abi: fakeAbi,
              address: fakeRx.contractAddress,
              constructorArgs: [
                viem.stringToHex('one', { size: 32 }),
                viem.stringToHex('two', { size: 32 }),
                { three: 'four' },
              ],
              contractName: undefined,
              deployTxnHash: fakeRx.transactionHash,
              deployedOn: 'contract.Woot',
              deployTxnBlockNumber: '0',
              deployTimestamp: '',
              linkedLibraries: {},
              sourceName: undefined,
              highlight: undefined,
              gasCost: '5678',
              gasUsed: 1234,
            },
          },
        });
      });

      it('deploys with default signer', async () => {
        const result = await action.exec(
          fakeRuntime,
          fakeCtx,
          {
            artifact: 'hello',
            args: [viem.stringToHex('one', { size: 32 }), viem.stringToHex('two', { size: 32 }), { three: 'four' }],
            salt: 'wohoo',
            value: '1234',
          },
          { name: 'hello', version: '1.0.0', currentLabel: 'contract.Woot' }
        );

        expect(result).toStrictEqual({
          contracts: {
            Woot: {
              abi: fakeAbi,
              address: fakeRx.contractAddress,
              constructorArgs: [
                viem.stringToHex('one', { size: 32 }),
                viem.stringToHex('two', { size: 32 }),
                { three: 'four' },
              ],
              contractName: undefined,
              deployTxnHash: fakeRx.transactionHash,
              deployedOn: 'contract.Woot',
              deployTxnBlockNumber: '0',
              deployTimestamp: '',
              linkedLibraries: {},
              sourceName: undefined,
              highlight: undefined,
              gasCost: '5678',
              gasUsed: 1234,
            },
          },
        });
      });
    });
  });
});
