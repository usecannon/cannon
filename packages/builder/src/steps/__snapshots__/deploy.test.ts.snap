// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`steps/deploy.ts exec() when create2 = true fails if contract already deployed 1`] = `"The contract at the create2 destination 0x3F9270CE7b8704E7BE0BfcA0EA8836f2B135a4ef is already deployed, but the Cannon state does not recognize that this contract has already been deployed. This typically indicates incorrect upgrade configuration. Please confirm if this contract should already be deployed or not, and if you want to continue the build as-is, add 'ifExists = "continue"' to the step definition"`;
